components/core/src/streaming_archive/writer/Archive.cpp

```diff
@@ -237,37 +235,56 @@ namespace streaming_archive { namespace writer {
         m_path.clear();
     }
 
-    File* Archive::create_file (const string& path, const group_id_t group_id, const boost::uuids::uuid& orig_file_id, size_t split_ix) {
+    void Archive::create_and_open_file (const string& path, const group_id_t group_id, const boost::uuids::uuid& orig_file_id, size_t split_ix) {
+        if (m_file != nullptr) {
+            throw OperationFailed(ErrorCode_NotReady, __FILENAME__, __LINE__);
+        }
         auto file = new File(m_uuid_generator(), orig_file_id, path, group_id, split_ix);
-        m_mutable_files.insert(file);
+        m_file = file;
+        m_file->open();
+    }
 
-        return file;
+    void Archive::close_file () {
+        m_file->close();
     }
 
-    void Archive::open_file (File& file) {
-        file.open();
+    bool Archive::is_file_open () {
+        return m_file->is_open();
     }
 
-    void Archive::close_file (File& file) {
-        file.close();
+    void Archive::change_ts_pattern (const TimestampPattern* pattern) {
+        m_file->change_ts_pattern(pattern);
     }
 
-    bool Archive::is_file_open (File& file) {
-        return file.is_open();
+    void Archive::append_log_id_to_segment(const logtype_dictionary_id_t log_id) {
+        if (m_file->has_ts_pattern()) {
+            m_logtype_ids_in_segment_for_files_with_timestamps.insert_id(log_id);
+        } else {
+            m_log_ids_without_timestamps_temp_holder.insert(log_id);
+        }
     }
 
-    void Archive::change_ts_pattern (File& file, const TimestampPattern* pattern) {
-        file.change_ts_pattern(pattern);
+    void Archive::append_var_ids_to_segment(const std::vector<variable_dictionary_id_t>& var_ids) {
+        if (m_file->has_ts_pattern()) {
+            for(auto id : var_ids){
+                m_var_ids_in_segment_for_files_with_timestamps.insert_id(id);
+            }
+        } else {
+            for(auto id : var_ids){
+                m_var_ids_without_timestamps_temp_holder.insert(id);
+            }
+        }
     }
 
-    void Archive::write_msg (File& file, epochtime_t timestamp, const string& message, size_t num_uncompressed_bytes) {
+    void Archive::write_msg (epochtime_t timestamp, const string& message, size_t num_uncompressed_bytes) {
         vector<encoded_variable_t> encoded_vars;
         vector<variable_dictionary_id_t> var_ids;
         EncodedVariableInterpreter::encode_and_add_to_dictionary(message, m_logtype_dict_entry, m_var_dict, encoded_vars, var_ids);
         logtype_dictionary_id_t logtype_id;
         m_logtype_dict.add_entry(m_logtype_dict_entry, logtype_id);
-
-        file.write_encoded_msg(timestamp, logtype_id, encoded_vars, var_ids, num_uncompressed_bytes);
+        append_var_ids_to_segment(var_ids);
```

Comment:
* Since these two methods are only used in this one place and they repeat the same if conditions, I would inline them here instead.
* Then to improve readability, I would:
  * add a comment at the top of the method `// Encode message and add components to dictionaries`;
  * add a newline after adding the logtype to the dictionary;
  * add a newline after writing the message;
  * add a comment before the inlined methods saying something like `// Update segment indices`